### ISSUE TITLE
False Positive: Expressions that already compute in a wider type are still misclassified as `int` overflow.

### ISSUE THREAD

**Issue by Carlson-JLQ** [google/error-prone#5638]
# False Positive: Expressions that already compute in a wider type are still misclassified as `int` overflow.

Version
2.48.0

## Checker
- Checker id: `IntLongMath`
- Checker description: Detects when an int-typed arithmetic expression may overflow before being assigned to a long-typed variable, field, or return value.

## Description of the false positive
`diff/output` marks 1 false positives for `IntLongMath`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The reports read more like over-eager matching than strong bug findings.

### `NegCase2.java`
Source: `output/codeql/IntMultToLong/src/main/java/scensct/core/neg/NegCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: Multiplication implicitly converted to same-sized or non-integral type should not be flagged as overflow before widening to larger integral type.

```java
// Multiplication implicitly converted to same-sized or non-integral type should not be flagged as overflow before widening to larger integral type.
package scensct.core.neg;

public class NegCase2 {
    public static void main(String[] args) {
        int x = 100;
        int y = 200;
        long sameSize = x * y; // Actually widening to long, but scenario example: same-sized conversion not possible for int to long? long is larger. Adjust to float.
        float toFloat = x * y; // Scenario 2: Implicit conversion to float (non-integral type)
        System.out.println(toFloat);
    }
}
```

## Cause analysis
The rule is documented as follows: Detects when an int-typed arithmetic expression may overflow before being assigned to a long-typed variable, field, or return value.

 The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `IntLongMath`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
